File: rust/src/bin/day4.rs
use std::cmp;
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let lines: Vec<String> = read_lines("../inputs/day4.txt")
        .unwrap()
        .map_while(Result::ok)
        .collect();
    part1(lines.clone());
    part2(lines.clone());
}
fn find(xi: i32, yi: i32, lines: Vec<String>) -> i32 {
    let mut c = 0;
    let directions = [
        (1i32, 0i32),
        (-1, 0),
        (0, 1),
        (0, -1),
        (1, 1),
        (1, -1),
        (-1, 1),
        (-1, -1),
    ];
    let max_y: i32 = lines.len() as i32 - 1;
    let max_x: i32 = lines[0].len() as i32 - 1;
    for (dx, dy) in directions {
        let mut sub_str = "X".to_string();
        let mut cx = xi;
        let mut cy = yi;
        for _ in 0..3 {
            cx += dx;
            cy += dy;
            if !((0i32 <= cx && cx <= max_x) && (0i32 <= cy && cy <= max_y)) {
                sub_str = "X".to_string();
                break;
            }
            sub_str.push(lines[cy as usize].chars().nth(cx as usize).unwrap());
        }
        if sub_str == "XMAS" {
            c += 1
        }
    }
    c
}
fn find2(xi: i32, yi: i32, lines: Vec<String>) -> i32 {
    let max_y: i32 = lines.len() as i32 - 1;
    let max_x: i32 = lines[0].len() as i32 - 1;

    if !(((0 <= (xi + 1)) && ((xi + 1) <= max_x)) && ((0 <= (xi - 1)) && ((xi - 1) <= max_x))) {
        return 0;
    }
    if !(((0 <= (yi + 1)) && ((yi + 1) <= max_y)) && ((0 <= (yi - 1)) && ((yi - 1) <= max_y))) {
        return 0;
    }

    let mut right_to_left = lines[(yi - 1) as usize]
        .chars()
        .nth((xi - 1) as usize)
        .unwrap()
        .to_string();
    right_to_left.push(
        lines[(yi + 1) as usize]
            .chars()
            .nth((xi + 1) as usize)
            .unwrap(),
    );
    let mut left_to_right = lines[(yi + 1) as usize]
        .chars()
        .nth((xi - 1) as usize)
        .unwrap()
        .to_string();
    left_to_right.push(
        lines[(yi - 1) as usize]
            .chars()
            .nth((xi + 1) as usize)
            .unwrap(),
    );
    if (right_to_left.eq("MS") || right_to_left.chars().rev().collect::<String>().eq("MS"))
        && (left_to_right.eq("MS") || left_to_right.chars().rev().collect::<String>().eq("MS"))
    {
        return 1;
    }
    0
}

fn part1(lines: Vec<String>) {
    let ans: i32 = lines
        .clone()
        .into_iter()
        .enumerate()
        .map(|(yi, line)| {
            line.chars()
                .enumerate()
                .map(|(xi, ch)| {
                    if ch == 'X' {
                        find(xi as i32, yi as i32, lines.clone())
                    } else {
                        0_i32
                    }
                })
                .sum::<i32>()
        })
        .sum();
    println!("pt1: {}", ans);
}
fn part2(lines: Vec<String>) {
    let ans: i32 = lines
        .clone()
        .into_iter()
        .enumerate()
        .map(|(yi, line)| {
            line.chars()
                .enumerate()
                .map(|(xi, ch)| {
                    if ch == 'A' {
                        find2(xi as i32, yi as i32, lines.clone())
                    } else {
                        0_i32
                    }
                })
                .sum::<i32>()
        })
        .sum();
    println!("pt2: {}", ans);
}
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

File: python/day04/day4.py
import re
import heapq
from collections import deque, defaultdict, Counter


input = open("../../inputs/day4.txt").readlines()


def find(xi, yi, grid):
    c = 0
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
    max_y = len(grid) - 1
    max_x = len(grid[0]) - 1
    for dx, dy in directions:
        sub_str = "X"
        cx = xi
        cy = yi
        for _ in range(3):
            cx = cx + dx
            cy = cy + dy
            if not (0 <= cx <= max_x and 0 <= cy <= max_y):
                sub_str = "X"
                break
            sub_str += grid[cy][cx]
        if sub_str == "XMAS":
            c += 1

    return c


def find2(xi, yi, grid):
    max_y = len(grid) - 1
    max_x = len(grid[0]) - 1
    if not ((0 <= xi + 1 <= max_x) and (0 <= xi - 1 <= max_x)):
        return 0
    if not ((0 <= yi + 1 <= max_y) and (0 <= yi - 1 <= max_y)):
        return 0

    right_to_left = grid[yi - 1][xi - 1] + grid[yi][xi] + grid[yi + 1][xi + 1]
    left_to_right = grid[yi + 1][xi - 1] + grid[yi][xi] + grid[yi - 1][xi + 1]
    if (right_to_left == "MAS" or right_to_left[::-1] == "MAS") and (
        left_to_right == "MAS" or left_to_right[::-1] == "MAS"
    ):
        return 1
    return 0


def part1(lines):
    ans = 0
    for yi, line in enumerate(lines):
        for xi, char in enumerate(line):
            if char == "X":
                ans += find(xi, yi, lines)
    return ans


def part2(lines):
    ans = 0
    for yi, line in enumerate(lines):
        for xi, char in enumerate(line):
            if char == "A":
                ans += find2(xi, yi, lines)
    return ans


print(part1(input))
print("***********part 2***************")
print(part2(input))
